src/transport/mqtt/connection.rs

```diff
@@ -0,0 +1,893 @@
+}
+
+impl InitError {
+    const fn client(ctx: &'static str) -> impl Fn(ClientError) -> InitError {
+        move |backtrace: ClientError| InitError::Client { backtrace, ctx }
+    }
+
+    const fn notice(ctx: &'static str) -> impl Fn(NoticeError) -> InitError {
+        move |backtrace: NoticeError| InitError::Notice { backtrace, ctx }
+    }
+}
+
+/// MQTT connection to Astarte that can be pulled to receive packets.
+#[derive(Debug)]
+pub(crate) struct MqttConnection {
+    connection: Connection,
+    /// Queue for the packet published while re/connecting.
+    buff: VecDeque<Publish>,
+    state: State,
+}
+
+impl MqttConnection {
+    pub(crate) fn new(
+        client: AsyncClient,
+        eventloop: EventLoop,
+        provider: TransportProvider,
+        state: impl Into<State>,
+    ) -> Self {
+        let connection = Connection {
+            client,
+            eventloop: SyncWrapper::new(eventloop),
+            provider,
+        };
+
+        Self {
+            connection,
+            buff: VecDeque::new(),
+            state: state.into(),
+        }
+    }
+
+    /// Iterate for the next publish event from the connection.
+    ///
+    /// Return [`None`] when disconnected.
+    pub(crate) async fn next_publish(&mut self) -> Option<Publish> {
+        // Check if there are backed up publishes.
+        if let Some(publish) = self.buff.pop_front() {
+            return Some(publish);
+        }
+
+        // Here we only get the connected state so we don't need to pass all the arguments, like the
+        // interfaces or the store.
+        let State::Connected(connected) = &mut self.state else {
+            return None;
+        };
+
+        loop {
+            match connected.poll(&mut self.connection).await {
+                Next::Same => {}
+                Next::Publish(publish) => return Some(publish),
+                Next::State(next) => {
+                    debug_assert!(!next.is_connected());
+
+                    self.state = next;
+
+                    return None;
+                }
+            }
+        }
+    }
+
+    /// Wait for the connection to be established to the Astarte.
+    ///
+    /// This function assumes that the connection is valid, but will handle disconnection and
+    /// reconnection automatically.
+    pub(crate) async fn wait_connack<S>(
+        client: AsyncClient,
+        eventloop: EventLoop,
+        provider: TransportProvider,
+        client_id: ClientId<&str>,
+        interfaces: &Interfaces,
+        store: &StoreWrapper<S>,
+    ) -> Result<Self, StoreError>
+    where
+        S: PropertyStore,
+    {
+        let mut mqtt_connection = Self::new(client, eventloop, provider, Connecting);
+
+        mqtt_connection
+            .reconnect(client_id, interfaces, store)
+            .await?;
+
+        Ok(mqtt_connection)
+    }
+
+    /// Connect to astarte, wait till the state is [`Connected`].
+    pub(crate) async fn reconnect<S>(
```

Comment:
Why did you call it `reconnect`? It seems more like something about the first connection

Reply:
Because we don't start in a disconnected state, so we have being disconnected and we are reconnecting. The first state is the `Connecting` state.

Reply:
Yeh, in an earlier version, the wait_connack and reconnect were two different functions now it makes more sens to call it `connect`